Introduce User Groups

lukasaric opened the pull request:
This PR introduces `user-group` term.
- Group is defined as a hierarchical structure which means it has sub-groups within.
- It adds the possibility to add users within groups
- It provides the possibility to change the `user` role within the group.
- Also it enables adding groups within the offerings (program/course)
- It sets access rights for `user-groups` and `members` within

TODO:

- [ ]  Group-based enrollments

lukasaric committed:
Basic server setup :rocket:

- Added group model
- Added group-user relationship
- Basic routing setup
- Basic controller CRUD setup

lukasaric committed:
Merge branch 'feature/sequelize-v6' into feature/introduce-user-groups

lukasaric committed:
Mount Group router on main router

lukasaric committed:
Client basic setup :surfer:

- Routing setup :eyes:
- Added group Table & modal

lukasaric committed:
GroupUser model options cleanup

lukasaric committed:
Add paramsParser util & prepared members route

lukasaric committed:
Avoid using composite key in group-user table

lukasaric committed:
Added Members table & dialog

lukasaric committed:
Dialog cleanup

lukasaric committed:
Merge branch 'feature/sequelize-v6' into feature/introduce-user-groups

lukasaric committed:
UI/UX alignments

lukasaric committed:
Tables & relations naming improvements :eyes:

lukasaric committed:
Set user-group filter in motion :rocket:

lukasaric committed:
:art:

lukasaric committed:
Added basic members listing :bowtie:

lukasaric committed:
Icons alignment :nail_care:

lukasaric committed:
Added group member creation

lukasaric committed:
Added update, remove & restore

lukasaric committed:
Merge branch 'feature/sequelize-v6' into feature/introduce-user-groups

lukasaric committed:
Merge branch 'develop' into feature/introduce-user-groups

lukasaric committed:
Updated migrations timestamps

lukasaric committed:
Role import fix

lukasaric committed:
Mounted user group member as user group sub-router

lukasaric committed:
Adapted similar UX flow as on programs

lukasaric committed:
Infer create & restore logic

lukasaric committed:
Removed user groups store module

lukasaric committed:
Added user full name & label virtual fields

lukasaric committed:
Added user scope

lukasaric committed:
Merge branch 'feature/enrollment-offerings' into feature/introduce-user-groups

lukasaric committed:
:wrench:

lukasaric committed:
Added offering-user-group relation

- Isolated offering route from enrollment
- Set offering-user-group router as offering sub-router
- Folder naming cleanup

lukasaric committed:
Implemented adding offering user groups on UI

lukasaric committed:
Add proper id to app bar dense prop

lukasaric committed:
Add router link per offering user group

lukasaric committed:
Fixed offering user groups filter

lukasaric committed:
Added basic sub-group setup

lukasaric committed:
User sub groups improvements

lukasaric committed:
Adjust user group listing behaviour

lukasaric committed:
Show all breadcrumb items user group ancestors

lukasaric committed:
breadcrumbs cleanup

lukasaric committed:
Added on fly fetch for breadcrumbs

lukasaric committed:
Set breadcrum items as router links

lukasaric committed:
Implement sub groups as data table :scroll:

lukasaric committed:
Reused common user-select on enrollment dialog

lukasaric committed:
UI/UX tweaks

lukasaric committed:
Used learner/user label instead of fullname util

lukasaric committed:
Controllers cleanup :nail_care:

lukasaric committed:
Infer offering & userGroupOffering client api
- Remove already added offering user groups in autocomplete

lukasaric committed:
Extract common user group selection & cleanup

lukasaric committed:
Added outside click on dialogs

lukasaric committed:
UI/UX tweaks

lukasaric committed:
Used user label on navbar menu tooltip

lukasaric committed:
More breadcrumbs enhancements

lukasaric committed:
:lipstick:

lukasaric committed:
:art:

lukasaric committed:
Icons & headings alignment :eyes:

lukasaric committed:
Better handled sub group data hydration

lukasaric committed:
Set user groups on offering route landing

lukasaric committed:
Separated public/protected routes on main router

lukasaric committed:
Detached member and user roles

lukasaric committed:
Rename offering user group migration

lukasaric committed:
Added basic user group member access rights

lukasaric committed:
Basic user group access rights :rocket:

lukasaric committed:
Group access rights cleanup

lukasaric committed:
:nail_care:

lukasaric committed:
Cleanup

lukasaric committed:
Allow self member deletion (leave group)

lukasaric committed:
User list controller cleanup

lukasaric committed:
Hydrate data on settings update

lukasaric committed:
:lipstick:

lukasaric committed:
Extract common toolbar css in main styles

lukasaric committed:
Event naming :eyes:

lukasaric committed:
Show open or restore

lukasaric committed:
Added user group cascade deletion & restore

lukasaric committed:
:art:

lukasaric committed:
Removed leftover getting single user group

- Parsed paranoid option in query processor

lukasaric committed:
Optimize paranoid in `parseOptions`

lukasaric committed:
Added filtering by member role :bowtie:

lukasaric committed:
:lipstick:

lukasaric committed:
Merge branch 'feature/enrollment-offerings' into feature/introduce-user-groups

lukasaric committed:
Replace transform with resursive function

lukasaric committed:
Resursive function cleanup :lipstick:

lukasaric committed:
Fixed express async err require order

lukasaric committed:
Merge branch 'feature/enrollment-offerings' into feature/introduce-user-groups

lukasaric committed:
App requires :wrench:

lukasaric committed:
Infered user group access check in one fn

lukasaric committed:
Reverted user group access changes

lukasaric committed:
User group access rights minor corrections

lukasaric committed:
Set learner as instructor after group creation

- Infered access rights

lukasaric committed:
On user group creation instructor setup polish

lukasaric committed:
User group member listing filter fix

lukasaric committed:
Extract parentId from body in access manager

lukasaric committed:
Minor user-groups breadcrumbs fix

lukasaric committed:
Removed unnecessary reverse on breadcrumbs

lukasaric committed:
Set first parent group admin

lukasaric committed:
Fixed reversed breadcrumb items args

lukasaric committed:
Removed unnecessary member group access mw

lukasaric committed:
Managed ancestors instructors access rights

underscope commented on server/common/database/migrations/20210201082605-user-group-member-relationship.js:
```diff
@@ -0,0 +1,45 @@
+    type: INTEGER,
+    primaryKey: true,
+    autoIncrement: true,
+    allowNull: false
```
Default with prev set?

lukasaric committed:
Merge branch 'develop' into feature/introduce-user-groups

underscope commented on server/common/database/migrations/20210201082605-user-group-member-relationship.js:
```diff
@@ -0,0 +1,45 @@
+const TABLE_NAME = 'user_group_member';
+
+exports.up = (qi, { DATE, ENUM, INTEGER }) => qi.createTable(TABLE_NAME, {
+  id: {
```
Why not having userId nad userGroupId as a composite?

underscope commented on server/common/database/migrations/20210201082605-user-group-member-relationship.js:
```diff
@@ -0,0 +1,45 @@
+
+const { MemberRole } = require('../../../../common/config');
+
+const TABLE_NAME = 'user_group_member';
```
Consider renaming to user_group_membership

underscope commented on server/common/database/migrations/20210201082605-user-group-member-relationship.js:
```diff
@@ -0,0 +1,45 @@
+    allowNull: false
+  },
+  role: {
+    type: ENUM(Object.values(MemberRole))
```
Migrations shouldn't rely on external objects (can be changed, should be immutable).

lukasaric committed:
Get user group id from req properly

underscope commented on common/config/index.js:
```diff
@@ -1,7 +1,9 @@
 module.exports = {
-  Role
+  Role,
+  MemberRole
```
UserGroupRole?

underscope commented on server/common/database/migrations/20210204150334-offering-user-group-relationship.js:
```diff
@@ -0,0 +1,40 @@
+'use strict';
```
use _ within filename for table name (apply to other migrations as well)

underscope commented on server/common/database/migrations/20210204150334-offering-user-group-relationship.js:
```diff
@@ -0,0 +1,40 @@
+    type: INTEGER,
+    primaryKey: true,
+    autoIncrement: true,
+    allowNull: false
```
Defaut with prev set attrs.

underscope commented on server/common/database/migrations/20210204150334-offering-user-group-relationship.js:
```diff
@@ -0,0 +1,40 @@
+'use strict';
+
+const TABLE_NAME = 'offering_user_group';
```
Should we relate group-based enrollments?

underscope commented on server/common/database/migrations/20210204150334-offering-user-group-relationship.js:
```diff
@@ -0,0 +1,40 @@
+const TABLE_NAME = 'offering_user_group';
+
+exports.up = (qi, { DATE, INTEGER }) => qi.createTable(TABLE_NAME, {
+  id: {
```
Do we need id? (Composite)

underscope commented on server/common/database/migrations/20210204150334-offering-user-group-relationship.js:
```diff
@@ -0,0 +1,40 @@
+    references: { model: 'user_group', key: 'id' },
+    allowNull: false
+  },
+  enrollmentOfferingId: {
```
can be shortened to offeringId

underscope commented on server/user-group/user-group.model.js:
```diff
@@ -0,0 +1,99 @@
+    });
+    this.belongsToMany(EnrollmentOffering, {
+      through: OfferingUserGroup,
+      foreignKey: { name: 'enrollmentOfferingId', field: 'enrollment_offering_id' }
```
Consider shortening to offeringId

underscope commented on server/user-group/user-group.model.js:
```diff
@@ -0,0 +1,99 @@
+
+  static hooks() {
+    return {
+      afterDestroy(userGroup) {
```
Move rather to service since this will need to do a cleanup on Enrollments and other things as well down the line.

underscope commented on server/user-group/user-group.model.js:
```diff
@@ -0,0 +1,99 @@
+    const attributes = ['id', 'parentId'];
+    const User = this.sequelize.model('User');
+    const include = [{ model: User, as: 'members', attributes: ['id'] }];
+    return this.findOne({ where, attributes, include });
```
findByPK 🤔 ?

underscope commented on server/user/user.controller.js:
```diff
@@ -27,15 +27,16 @@ const inputAttrs = ['email', 'role', 'firstName', 'lastName'];
-    .then(({ rows, count }) => {
-      return res.jsend.success({ items: map(rows, 'profile'), total: count });
-    });
+  if (userIds) where[Op.not] = { id: userIds };
```
Not sure if it makes sense that sending userIds excludes them.

underscope commented on server/offering-user-group/index.js:
```diff
@@ -0,0 +1,27 @@
+const { OfferingUserGroup } = require('../common/database');
+
+router
+  .param('userGroupId', getUserGroup)
```
just groupId / id within this context ?

underscope commented on server/user-group-member/index.js:
```diff
@@ -0,0 +1,35 @@
+const router = require('express').Router();
+
+router
+  .param('memberId', getMember)
```
userId might be more intuitive param name

underscope reviewed (commented):
Here are a couple of comments for the sections I have reviewed.

lukasaric committed:
Initialized basic group member enrollment hooks

lukasaric committed:
Managed enrollments on member restore

lukasaric committed:
Router param naming polish

lukasaric committed:
Used `findByPk` instead :eyes:

lukasaric committed:
Naming cleanup :lipstick:

- User group roles naming polish
- Hyphened table names in migration file names
- Shorthened associations names

lukasaric committed:
Removed unnecessary allow null prop

lukasaric committed:
Extracted hooks as separate files :rocket:

- Changed hooks db registration
- Extracted user group & group-membership hooks logic

lukasaric committed:
User group membership hooks cleanup :nail_care:

lukasaric committed:
Remove/restore enrollment on group remove/restore

lukasaric committed:
Nomenclature & primary keys improvements

lukasaric committed:
Membership access rights improvements

lukasaric committed:
Membership mws naming cleanup

lukasaric committed:
Removed table actions val

lukasaric committed:
Fixed offering user group foreign key associations

lukasaric committed:
Add basic offering user group destroy hook

- Managed groups descendants
- Handled enrollments after group is detached

lukasaric committed:
Removed membership on userId update logic

lukasaric committed:
Unified user-group & membership :rocket:

- Replaced enrollment service with hooks
- Integrated transactions :eyes:
- Merged both routers

lukasaric committed:
Send prev deleted_at as param directly

lukasaric committed:
Added cascade deletion within user group migration

lukasaric committed:
Unified offering & offering-user-group :surfer:

- Merged both routers
- Removed hooks, moved logic to enrollment service

lukasaric committed:
Handled offering user group enrollment :rocket:

lukasaric committed:
Fixed wrong where param

lukasaric committed:
Exclude duplicate users from other groups

lukasaric committed:
Remove & restore traversing improvement

lukasaric committed:
Enrollment service enhancement :rocket:

- Handled all unenroll logic better
- Minimized methods
- Code cleanup

lukasaric committed:
Removed unnecessary class based approach

lukasaric committed:
Removed offering ids helper fn

lukasaric committed:
Add todo comment on `getDescendants` method

lukasaric committed:
Begin transaction within the controllers

- Pass transactions in enrollment service

lukasaric committed:
Implement CTE for getting group descendents

lukasaric committed:
Indent raw query

lukasaric committed:
Applied proper descendant method in remove
